### PR TITLE
374 Import Manager Update

### LINKED ISSUE #374: Update the tally system to handle new Ballot types more generically
### User story
---
Currently the tally system supports 3 ballot types and can support new ballot types but with the following constraints:

  - Updating of the initial tally creation files in a particular format to take into account the new ballots.
  - Refactoring the tally app to accommodate the new ballot types.
  
This is not the desired workflow as it is prone to human errors and a longer feedback loop between the client and the devs. That said, we need a more generic solution that can configurable per deployment and handle a variety of Ballot types without the above mentioned constraints.

### Proposed solution
---
- A Race Django model will be used to hold Race types. This model will be populated by a configurable list of Race types from the Django settings so that in case we have new one's, they will just be updated on deployment.
- Race creation will upon a deployment via a Django management command executed during the deployment.
- The configurable list of Races will consist of:
   - Race Type
   - Race Code
   - Ballot Name. **NB** This should be similar to the Ballot Number column name in the sub constituency file.
- The Ballot model will be updated to have a Many to 1 relationship with the Race model for assignment of Races to Ballots.
   - With this change the current  `race_type` property on Ballot Model will be retrieved from the Race Model hence this property will be removed from the model.
-  Refactor the tally app to use Race Model where it's currently using the Races Enum.

### Acceptance Criteria
---
- [ ] Races should be configurable from a config file. Django settings.
- [ ] Configured Races should be created/updated through a deployment.

### Task check list
---
- [ ] Create a Race Model with the following properties:
   - type = string
   - code = number
   - ballot_name = string
- [ ] Create an Environment variable called Races which will be an Array of race type dictionaries with the following properties:
    - type
   - code
   - ballot_name
- [ ] Create a `create_races` Django management command that will pick the current configure RACES and create Race objects.
- [ ] Update the tallyho ansible role to always to always run the `create_races` Django management command. The command can be added to this [list](https://github.com/onaio/ansible-tally-ho/blob/d77a6fa2cc9ac062f9a19d62018ae2b7c884008b/defaults/main.yml#L143).
- [ ] Refactor the tally app to use Race Model where it's currently using the Races Enum.

### PR DESCRIPTION
**Pull request by JohnMwashuma**
Fixes #374 

### Task check list
---
- [x] Make race types dynamic, replacing them with electrol races.
- [x] Add ballots and sub con ballots files to the import work flow.
- [x] Refactor tally creation logic into separate files based on the type of file being processed and objects being created.
- [x] Bulk create tally objects to speed up the tally creation process.
- [x] Use celery to make the import process async, cache elements processed in memcache for the client side to pull the progress.
- [x] Refactor all references to the Race Type enum to use ballot electrol races election level and sub race.
- [x] Fix existing failing tests.
- [x] Add tests for the suggested changes.

### Related to
---
- https://github.com/onaio/ansible-tally-ho/pull/17
- https://github.com/onaio/infrastructure/pull/7122

### CODE REVIEW

File: tally_ho/apps/tally/views/reports/administrative_areas_reports.py [hunk 1]
```diff
@@ -672,175 +717,245 @@
...
-                    When(
-                        result_form__reconciliationform__isnull=False,
-                        then=F(reconform_num_valid_votes)
-                    ), default=V(0))).distinct()
+            )
+
+    return qs
+
+
+def export_results_queryset(
```
[the remark this answers is not in the document]
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/reports/administrative_areas_reports.py:766]: Tracked here https://github.com/onaio/tally-ho/issues/414

File: tally_ho/apps/tally/views/super_admin.py [hunk 2]
```diff
@@ -1205,21 +1229,182 @@
...
+        tally_id = self.kwargs.get('tally_id')
+        context = super(EditElectrolRaceView, self).get_context_data(**kwargs)
+        context['id'] = self.kwargs.get('id')
+        context['tally_id'] = tally_id
+        context['is_active'] = self.object.active
+        context['comments'] = self.object.comments.filter(tally__id=tally_id)
+
+        return context
+
+    def get_object(self):
```
[the remark this answers is not in the document]
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/super_admin.py:1308]: Dismissed

File: tally_ho/apps/tally/forms/edit_electrol_race_form.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        label=_("Add new comments"),
+        widget=forms.Textarea(attrs={'cols': 80, 'rows': 5}),
+    )
+
+    def __init__(self, *args, **kwargs):
+        super(EditElectrolRaceForm, self).__init__(*args, **kwargs)
+        if self.instance.active:
+            self.fields.pop('disable_reason')
+
+    def save(self):
```
[the remark this answers is not in the document]
Comment (JohnMwashuma) [reply tally_ho/apps/tally/forms/edit_electrol_race_form.py:40]: Dismissed

File: tally_ho/apps/tally/views/super_admin.py [hunk 4]
```diff
@@ -1205,21 +1229,182 @@
...
+        tally_id = self.kwargs.get('tally_id')
+        id = self.kwargs.get('id')
```
[the remark this answers is not in the document]
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/super_admin.py]: Fixed

File: tally_ho/apps/tally/views/super_admin.py [hunk 5]
```diff
@@ -1205,21 +1229,182 @@
...
+
+        return get_object_or_404(ElectrolRace, tally__id=tally_id, id=id)
+
+    def get_success_url(self):
+        tally_id = self.kwargs.get('tally_id')
+        id = self.kwargs.get('id')
```
[the remark this answers is not in the document]
Comment (JohnMwashuma) [reply tally_ho/apps/tally/views/super_admin.py]: Fixed

**Review by lincmba — APPROVED** [review]

**Pull request merged**